### PR TITLE
JDK-8293156: Dcmd VM.classloaders fails to print the full hierarchy

### PR DESCRIPTION
**Pull request by tstuefe**
Fixes a bug in the `VM.classloaders` jcmd that causes class loaders to be omitted from the output if a parent class loader never loaded any class and therefore had no associated DCmd.

The fix changes the command to not rely on the existence of a CLD structure for the loader; instead, all information (loader class name, loader name, etc) is pulled via the loader oop, which has to be always there unless its the bootstrap loader.

Also, the tests were expanded to test the display of empty loaders and empty parent loaders.

Thanks to @dholmes-ora for finding this bug.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293156](https://bugs.openjdk.org/browse/JDK-8293156): Dcmd VM.classloaders fails to print the full hierarchy

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10312/head:pull/10312` \
`$ git checkout pull/10312`

Update a local copy of the PR: \
`$ git checkout pull/10312` \
`$ git pull https://git.openjdk.org/jdk pull/10312/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10312`

View PR using the GUI difftool: \
`$ git pr show -t 10312`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10312.diff">https://git.openjdk.org/jdk/pull/10312.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp [hunk 1]
```diff
@@ -154,33 +154,59 @@ class LoaderTreeNode : public ResourceObj {
...
+
+  bool is_bootstrap() const {
+    if (_loader_oop == NULL) {
+      assert(_cld != NULL && _cld->is_boot_class_loader_data(), "bootstrap loader must have CLD");
+      return true;
+    }
+    return false;
+  }
+
   void print_with_childs(outputStream* st, BranchTracker& branchtracker,
```
Comment (plummercj) [inline src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:187]: I know this isn't part of your changes, but the plural of child is children, not childs. However, I think child_nodes would probably be best here.

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp [hunk 2]
```diff
@@ -287,8 +315,9 @@ class LoaderTreeNode : public ResourceObj {
   // and they have the same name or no name (note: leaf check is done by caller).
   bool can_fold_into(LoaderTreeNode* target_node) const {
     assert(is_leaf() && target_node->is_leaf(), "must be leaf");
-    return _cld->class_loader_klass() == target_node->_cld->class_loader_klass() &&
-           _cld->name() == target_node->_cld->name();
+    return (loader_klass() != NULL) && (loader_klass() == target_node->loader_klass()) &&
+           ((loader_name_oop() == NULL && target_node->loader_name_oop() == NULL) ||
+             (::strcmp(loader_name(), target_node->loader_name()) == 0));
```
Comment (plummercj) [inline src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:320]: This is hard to parse. I think using if/else statements would be cleaner.

File: test/hotspot/jtreg/serviceability/dcmd/vm/ClassLoaderHierarchyTest.java [hunk 3]
```diff
@@ -60,39 +73,77 @@ public class ClassLoaderHierarchyTest {
...
-        output.shouldMatch(".*TestClassLoader");
-        output.shouldMatch("Kevin.*TestClassLoader");
-        output.shouldMatch("Bill.*TestClassLoader");
+        // (A)
+        output.shouldContain("+-- <bootstrap>");
+        output.shouldContain("      +-- \"platform\", jdk.internal.loader.ClassLoaders$PlatformClassLoader");
+        output.shouldContain("      |     +-- \"app\", jdk.internal.loader.ClassLoaders$AppClassLoader");
+        output.shouldContain("      +-- \"Kevin\", ClassLoaderHierarchyTest$TestClassLoader");
+        output.shouldContain("      +-- ClassLoaderHierarchyTest$TestClassLoader");
+        output.shouldContain("      |     +-- \"Bill\", ClassLoaderHierarchyTest$TestClassLoader");
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/dcmd/vm/ClassLoaderHierarchyTest.java:124]: Can you test using one long string, and do the same for each section below? I think embedding \\R as the newline character should work on all platforms. If you run into unforeseen issues, then don't worry about.

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp]: Fixed

File: test/hotspot/jtreg/serviceability/dcmd/vm/ClassLoaderHierarchyTest.java (hunk 3 above)
Comment (tstuefe) [reply test/hotspot/jtreg/serviceability/dcmd/vm/ClassLoaderHierarchyTest.java:124]: Fixed

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp]: Fixed

**Comment by tstuefe** [comment]
Hi @plummercj , thanks for your Review. I worked most of your your feedback.

.. Thomas

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp (hunk 1 above)
Comment (plummercj) [reply src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp]: It looks like there are a couple more references to "childs" still present.

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp [hunk 4]
```diff
@@ -154,33 +154,59 @@ class LoaderTreeNode : public ResourceObj {
...
       st->print(" <bootstrap>");
     } else {
-      assert(!_cld->has_class_mirror_holder(), "_cld must be the primary cld");
-      if (loader_name != NULL) {
-        st->print(" \"%s\",", loader_name->as_C_string());
+      if (the_loader_name[0] != '\0') {
+        st->print(" \"%s\",", the_loader_name);
       }
-      st->print(" %s", loader_klass != NULL ? loader_klass->external_name() : "??");
+      st->print(" %s", the_loader_class_name);
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:209]: `the_loader_class_name` could be NULL here. You need to change `loader_class_name()` t return "??" like the original code uses.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Thomas,

Generally this seems fine - but one issue below.

Thanks.

File: src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp (hunk 4 above)
Comment (tstuefe) [reply src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:209]: Good catch. We never encounter this because Klass can only be null for the bootstrap loader, which is handled in a different branch. I'll fix it.

**Comment by tstuefe** [comment]
> Hi @plummercj , thanks for your Review. I worked most of your your feedback.
> 
> .. Thomas

**Comment by tstuefe** [comment]
- Fixed all remaining occurrences of "childs"
- Fixed possible segfault when printing class name for bootstrap loader
- Fixed GHAs (apparently we don't have a MainWrapper class when executing jtreg in GHAs, so I cannot parse for it)

**Review by plummercj — APPROVED** [review]

**Comment by tstuefe** [comment]
Thanks @plummercj and @dholmes-ora .

/integrate
